File: src/main/java/icu/samnyan/aqua/sega/general/filter/CompressionFilter.java
package icu.samnyan.aqua.sega.general.filter;

import icu.samnyan.aqua.sega.util.Compression;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * @author samnyan (privateamusement@protonmail.com)
 */
@Component
public class CompressionFilter extends OncePerRequestFilter {

    private static final Logger logger = LoggerFactory.getLogger(CompressionFilter.class);
    private final List<String> filterList;

    public CompressionFilter() {
        filterList = new ArrayList<>();
        filterList.add("/ChuniServlet");
        filterList.add("/OngekiServlet");
        filterList.add("/MaimaiServlet");
        filterList.add("/Maimai2Servlet");
        filterList.add("/ChusanServlet");
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {

        logger.debug("Do compress filter");
        String encoding = request.getHeader("content-encoding");
        byte[] reqSrc = request.getInputStream().readAllBytes();

        byte[] reqResult;
        if (encoding != null && encoding.equals("deflate")) {
            logger.debug("Request length (compressed): {}", reqSrc.length);
            reqResult = Compression.decompress(reqSrc);
            logger.debug("Request length (decompressed): {}", reqResult.length);
        } else {
            reqResult = reqSrc;
        }

        CompressRequestWrapper requestWrapper = new CompressRequestWrapper(request, reqResult);
        CompressResponseWrapper responseWrapper = new CompressResponseWrapper(response);

        filterChain.doFilter(requestWrapper, responseWrapper);

        byte[] respSrc = responseWrapper.toByteArray();
        logger.debug("Response length (uncompressed): {}", respSrc.length);
        byte[] respResult = Compression.compress(respSrc);
        logger.debug("Response length (compressed): {}", respResult.length);


        response.setContentLength(respResult.length);
        response.setContentType("application/json; charset=utf-8");
        response.addHeader("Content-Encoding", "deflate");

        response.getOutputStream().write(respResult);

    }

    @Override
    protected boolean shouldNotFilter(HttpServletRequest request) {
        String path = request.getServletPath();
        boolean notFilter = true;
        for (String prefix : filterList) {
            if (path.startsWith(prefix)) {
                notFilter = false;
                break;
            }
        }
        return notFilter;
    }
}

File: src/main/java/icu/samnyan/aqua/sega/general/filter/CompressionFilter.kt
package icu.samnyan.aqua.sega.general.filter

import ext.logger
import icu.samnyan.aqua.sega.util.ZLib
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.eclipse.jetty.io.EofException
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import org.springframework.web.util.ContentCachingResponseWrapper
import java.util.*


/**
 * @author samnyan (privateamusement@protonmail.com)
 */
@Component
class CompressionFilter : OncePerRequestFilter() {
    companion object {
        val logger = logger()
        val b64d = Base64.getMimeDecoder()
        val b64e = Base64.getMimeEncoder()
    }

    override fun doFilterInternal(req: HttpServletRequest, resp: HttpServletResponse, chain: FilterChain) {
        val isDeflate = req.getHeader("content-encoding") == "deflate"
        val isDfi = req.getHeader("pragma") == "DFI"

        // Decode input
        val reqSrc = req.inputStream.readAllBytes().let {
            if (isDeflate) ZLib.decompress(it)
            else if (isDfi) ZLib.decompress(b64d.decode(it))
            else it
        }

        // Handle request
        val result = ContentCachingResponseWrapper(resp).run {
            chain.doFilter(CompressRequestWrapper(req, reqSrc), this)
            ZLib.compress(contentAsByteArray).let { if (isDfi) b64e.encode(it) else it }
        }

        // Write response
        resp.setContentLength(result.size)
        if (isDfi) resp.setHeader("pragma", "DFI")
        if (isDeflate) {
            resp.contentType = "application/json; charset=utf-8"
            resp.setHeader("content-encoding", "deflate")
        }

        try {
            resp.outputStream.use { it.write(result); it.flush() }
        } catch (e: EofException) {
            logger.warn("- EOF: Client closed connection when writing result")
        }
    }

    /**
     * Filter games that are not diva
     */
    override fun shouldNotFilter(req: HttpServletRequest) =
        !(req.servletPath.startsWith("/g/") && !req.servletPath.startsWith("/g/wacca"))
}
